cidc_api/utils/migrations.py

```diff
@@ -0,0 +1,159 @@
+
+
+class PieceOfWork(NamedTuple):
+    do: Callable[[], None]
+    undo: Callable[[], None]
+
+
+class RollbackableQueue:
+    """A collection of reversible pieces-of-work."""
+
+    def __init__(self):
+        self.tasks = []
+
+    def schedule(self, task: PieceOfWork):
+        """Add a task to the task queue."""
+        self.tasks.append(task)
+
+    def run_all(self):
+        """
+        Attempt to run all tasks in the queue, rolling back
+        successfully completed tasks if a subsequent task fails.
+        """
+        for i, task in enumerate(self.tasks):
+            try:
+                task.do()
+            except:
+                for done_task in self.tasks[:i]:
+                    done_task.undo()
+                raise
+
+
+def run_metadata_migration(metadata_migration: Callable[[dict], MigrationResult]):
+    """Migrate trial metadata, upload job patches, and downloadable files according to `metadata_migration`"""
+    session = Session(bind=op.get_bind())
+
+    # Initialize an empty list of pieces of work to do on GCS resources
+    gcs_tasks = RollbackableQueue()
+
+    # Migrate all trial records
+    trials: List[TrialMetadata] = session.query(TrialMetadata).with_for_update().all()
+    for trial in trials:
+        migration = metadata_migration(trial.metadata_json)
+
+        # Update the trial metadata object
+        trial.metadata_json = migration.result
+
+        # Update the relevant downloadable files and GCS objects
+        for old_gcs_uri, artifact in migration.file_updates.items():
+            # Update the downloadable file associated with this blob
+            df: DownloadableFiles = (
+                session.query(DownloadableFiles)
+                .filter_by(object_url=old_gcs_uri)
+                .with_for_update()
+                .one()
+            )
+            for column, value in artifact.items():
+                if hasattr(df, column):
+                    setattr(df, column, value)
+
+            # Regenerate additional metadata from the migrated clinical trial
+            # metadata object.
+            df.additional_metadata = _get_uuid_info(
+                migration.result, artifact["upload_placeholder"]
+            )[1]
+
+            # If the GCS URI has changed, rename the blob
+            new_gcs_uri = artifact["object_url"]
+            if old_gcs_uri != new_gcs_uri:
+                renamer = PieceOfWork(
+                    lambda: rename_gcs_blob(
+                        GOOGLE_DATA_BUCKET, old_gcs_uri, new_gcs_uri
+                    ),
+                    lambda: rename_gcs_blob(
+                        GOOGLE_DATA_BUCKET, new_gcs_uri, old_gcs_uri
+                    ),
+                )
+                gcs_tasks.schedule(renamer)
+
+    # Migrate all assay upload successes
+    successful_assay_uploads: List[AssayUploads] = session.query(
+        AssayUploads
+    ).filter_by(status=AssayUploadStatus.MERGE_COMPLETED.value).with_for_update().all()
```

Comment:
btw, same `with_for_update().all()` here doesn't look that concerning as the first one - here we're updating only  `MERGE_COMPLETED` assay uploads, that are probably not in the use by any other process, afaik.